--- ../build-GNU-4.9.3-64/pyplusplus/pyplusplus/module_builder/boost_python_builder.py	2015-10-27 13:32:49.000000000 +0100
+++ pyplusplus/module_builder/boost_python_builder.py	2015-10-27 13:36:45.000000000 +0100
@@ -44,7 +44,9 @@ class builder_t(module_builder.module_bu
                   , cflags=""
                   , encoding='ascii'
                   , compiler=None
-                  , gccxml_config=None):
+                  , compiler_path=None
+                  , gccxml_config=None
+                  , caster='gccxml' ):
         """
         :param files: list of files, declarations from them you want to export
         :type files: list of strings or :class:`parser.file_configuration_t` instances
@@ -72,7 +74,9 @@ class builder_t(module_builder.module_bu
         module_builder.module_builder_t.__init__( self, global_ns=None, encoding=encoding )
 
         if not gccxml_config:
-            gccxml_config = parser.gccxml_configuration_t( gccxml_path=gccxml_path
+            gccxml_config = parser.xml_generator_configuration_t(  
+                                            xml_generator_path=gccxml_path,
+                                            xml_generator=caster
                                              , working_directory=working_directory
                                              , include_paths=include_paths
                                              , define_symbols=define_symbols
@@ -80,7 +84,9 @@ class builder_t(module_builder.module_bu
                                              , start_with_declarations=start_with_declarations
                                              , ignore_gccxml_output=ignore_gccxml_output
                                              , cflags=cflags
-                                             , compiler=compiler)
+                                             , compiler=compiler
+                                             , compiler_path=compiler_path
+                                             )
 
         #may be in future I will add those directories to user_defined_directories to self.__code_creator.
         self.__parsed_files = list(map( pygccxml_utils.normalize_path
