diff -r d3879bfb9f12 indexing_suite_v2/indexing_suite/slice.hpp
--- a/indexing_suite_v2/indexing_suite/slice.hpp	Fri May 25 12:40:34 2018 -0500
+++ b/indexing_suite_v2/indexing_suite/slice.hpp	Wed Jul 18 20:28:26 2018 +0200
@@ -92,9 +92,10 @@
             PyErr_SetString (PyExc_ValueError, "slice step cannot be zero");
             boost::python::throw_error_already_set ();
         }
-
-        m_start = std::max (static_cast<index_type> (0), std::min (length, m_start));
-        m_stop = std::max (static_cast<index_type> (0), std::min (length, m_stop));
+        
+	// GetIndices(..) should ensure correct m_start and m_stop, the following will not work for slice(::-1)
+	//m_start = std::max (static_cast<index_type> (0), std::min (length, m_start));
+        //m_stop = std::max (static_cast<index_type> (0), std::min (length, m_stop));
         m_direction = (m_step > 0) ? 1 : -1;
     }
 
